### ISSUE TITLE
Code for BCH expansion

### ISSUE THREAD

**Issue by babbush** [quantumlib/OpenFermion#37]
Several times now I have wanted to implement code which performs a BCH expansion on operator data structures in OpenFermion such as QubitOperator, FermionOperator, and sparse matrices. I started to write some code for this:

```python

def bch_expand(x, y, order):
    """Compute log[e^x e^y] using the Baker-Campbell-Hausdorff formula
    Args:
        x: An operator for which multiplication and addition are supported.
            For instance, a QubitOperator, FermionOperator or numpy array.
        y: The same type as x.
        order(int): The order to truncate the BCH expansions. Currently
            function goes up to only third order.

    Returns:
        z: The truncated BCH operator.

    Raises:
        ValueError: operator x is not same type as operator y.
        ValueError: invalid order parameter.
        ValueError: order exceeds maximum order supported.
    """
    # Initialize.
    max_order = 4
    if type(x) != type(y):
        raise ValueError('Operator x is not same type as operator y.')
    elif (not isinstance(order, int)) or order < 0:
        raise ValueError('Invalid order parameter.')

    # Zeroth order.
    z = x + y

    # First order.
    if order > 0:
        z += commutator(x, y) / 2.

    # Second order.
    if order > 1:
        z += commutator(x, commutator(x, y)) / 12.
        z += commutator(y, commutator(y, x)) / 12.

    # Third order.
    if order > 2:
        z -= commutator(y, commutator(x, commutator(x, y))) / 24.

    # Fourth order.
    if order > 3:
        z -= commutator(
            y, commutator(y, commutator(y, commutator(y, x)))) / 720.
        z -= commutator(
            x, commutator(x, commutator(x, commutator(x, y)))) / 720.
        z += commutator(
            x, commutator(y, commutator(y, commutator(y, x)))) / 360.
        z += commutator(
            y, commutator(x, commutator(x, commutator(x, y)))) / 360.
        z += commutator(
            y, commutator(x, commutator(y, commutator(x, y)))) / 120.
        z += commutator(
            x, commutator(y, commutator(x, commutator(y, x)))) / 120.

    # Return.
    if order > max_order:
        raise ValueError('Order exceeds maximum order supported.')
    else:
        return z
```

But I am thinking that what we really want to do is to  implement the [form derived by Dynkin in 1947](https://en.wikipedia.org/wiki/Baker%E2%80%93Campbell%E2%80%93Hausdorff_formula#An_explicit_Baker.E2.80.93Campbell.E2.80.93Hausdorff_formula) which would enable us to compute the series to arbitrary order without hard coding expression. Coding this up is a little involved but it is a good starter project for somebody. Some thought should also go into optimizing the performance of such code and also making good tests.

**Comment by babbush** [comment]
Yet another issue to be fixed is why GitHub is randomly choosing not to format the top of my function correctly in the post above.

**Comment by DanielSank** [comment]
@babbush The formatting was messed up because you didn't use github's formatting ;-)

To do a code block (in python), you do this:

    ```python

    def foo(bar):
        """blah blah

        """
    ```

**Comment by DanielSank** [comment]
I fixed the formatting in the first comment.

**Comment by babbush** [comment]
Nice work. I will now delete these comments to save face...

**Comment by DanielSank** [comment]
@babbush Face is overrated.

**Comment by babbush** [comment]
Give vanity a chance.

**Comment by jdaaph** [comment]
I would love to contribute, can you assign it to me ~^~

**Comment by babbush** [comment]
Sure, we'd love to have you contribute! I can only use GitHub's "assign" on you if you accept the invitation I just sent to be an official collaborator of the project. But consider yourself assigned!

**Comment by jdaaph** [comment]
Some quick updates on the implementation: I've worked with my theorist friend to find out an efficient implementation for Dynkin's formula. I'm in the process of writing up the procedure in a short paper on arxiv and will refer to it in the implementation.

**Comment by babbush** [comment]
That's sounds awesome! I'm looking forward to it!

**Comment by jdaaph** [comment]
Hi, our code is finished, however it seems I don't have the permission to push my local branch to remote, please check and I'll push it tomorrow, thanks!

**Comment by babbush** [comment]
Indeed, nobody (not even me) has permission to push directly to the remote. Instead, you should open a pull request. See https://help.github.com/articles/about-pull-requests/ for details about doing this. Also, you will need to make sure you've sign the Google CLA (https://cla.developers.google.com/clas).

**Comment by babbush** [comment]
Closed by #114
